cirq/contrib/quimb/mps_simulator.py

```diff
@@ -235,83 +238,155 @@ def sample(
 class MPSState:
     """A state of the MPS simulation."""
 
-    def __init__(self, qubit_map, initial_state=0):
+    def __init__(self, qubit_map, rsum2_cutoff, initial_state=0):
         self.qubit_map = qubit_map
         self.M = []
-        for qubit in qubit_map.keys():
+
+        # The order of the qubits matters, because the state |01> is different from |10>. Since
+        # Quimb uses strings to name tensor indices, we want to be able to sort them too. If we are
+        # working with, say, 123 qubits then we want qubit 3 to come before qubit 100, but then
+        # we want write the string '003' which comes before '100' in lexicographic order. The code
+        # below is just simple string formatting.
+        max_num_digits = len('{}'.format(max(qubit_map.values())))
+        self.format_i = 'i_{{:0{}}}'.format(max_num_digits)
+        self.format_mu = 'mu_{{:0{}}}_{{:0{}}}'.format(max_num_digits, max_num_digits)
+
+        # TODO(tonybruguier): Refactor out so that the code below can also be used by
+        # circuit_to_tensors in cirq.contrib.quimb.state_vector.
+        for qubit in reversed(list(qubit_map.keys())):
             d = qubit.dimension
-            x = np.zeros(
-                (
-                    1,
-                    1,
-                    d,
-                )
-            )
-            x[0, 0, (initial_state % d)] = 1.0
-            self.M.append(x)
+            x = np.zeros(d)
+            x[initial_state % d] = 1.0
+
+            i = qubit_map[qubit]
+            self.M.append(qtn.Tensor(x, inds=(self.i_str(i),)))
             initial_state = initial_state // d
         self.M = self.M[::-1]
-        self.threshold = 1e-3
+        self.rsum2_cutoff = rsum2_cutoff
+        self.num_svd_splits = 0
+
+    def i_str(self, i):
+        return self.format_i.format(i)
+
+    def mu_str(self, i, j):
+        # By convention, the lower index is always the first.
+        smallest = min(i, j)
+        largest = max(i, j)
+        return self.format_mu.format(smallest, largest)
 
     def __str__(self) -> str:
-        return str(self.M)
+        return str([x.data for x in self.M])
 
     def _value_equality_values_(self) -> Any:
-        return self.qubit_map, self.M, self.threshold
+        return self.qubit_map, self.M, self.rsum2_cutoff
 
     def copy(self) -> 'MPSState':
-        state = MPSState(self.qubit_map)
+        state = MPSState(self.qubit_map, self.rsum2_cutoff)
         state.M = [x.copy() for x in self.M]
-        state.threshold = self.threshold
+        state.num_svd_splits = self.num_svd_splits
         return state
 
+    def _sum_up(self, skip_tracing_out_for_qubits):
+        end_bras = []
+        if skip_tracing_out_for_qubits:
+            for qubit, i in self.qubit_map.items():
+                if i not in skip_tracing_out_for_qubits:
+                    d = qubit.dimension
+                    data = [math.sqrt(d)] * d
```

Comment:
Why is this sqrt(d)? That means `end_bra` isn't normalized! Even if it was something like [1/sqrt(2), 1/sqrt(2)] I don't know why we'd want to "cap" the circuit with |+> state end bras. Maybe this is described in the paper?

Reply:
That's because I was wrong. I was assuming the state vector for the measured qubit was pure, which is not always the case. I was just summing out the other indices with a |+> that I had rescaled so that probs would add up to 1.0. However, I had only checked that the probs add up to 1.0 for one case, but it wasn't always the case.

I never caught the issue because all my testing had either pure states, or states |0> or |1>. In addition, I was always force-renormalizing to sum of probs to 1.0, even if they were very far from 1.0.

Instead, what I should have done is compute the partial trace, which is what I do now. I added a test that makes sure the partial trace is the same, and I tried to have a more coverage. In addition, I now have a tolerance for the sum of the probs, making sure they are close to 1.0 (that test was failing before my new code).

Thanks for catching this and sorry about this mistake.